### PR TITLE
otelzap: Add skeleton and config options 

### PR DESCRIPTION
**Pull request by khushijain21**
Part of  https://github.com/open-telemetry/opentelemetry-go-contrib/issues/5191

Pre-work https://github.com/open-telemetry/opentelemetry-go-contrib/pull/5279

### CODE REVIEW

**Comment by pellared** [comment]
Can you please outline an implementation plan in https://github.com/open-telemetry/opentelemetry-go-contrib/issues/5191?

File: bridges/otelzap/config.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package otelzap // import "otelzap"
```
Comment (pellared) [inline bridges/otelzap/config.go:4]: Have you added ` // import "otelzap"` manually? The import path should be `go.opentelemetry.io/contrib/bridges/otelzap` and should be generated automatically by `make`.

File: bridges/otelzap/config.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	if c.scope.Version != "" {
+		opts = append(opts, log.WithInstrumentationVersion(c.scope.Version))
+	}
+	if c.scope.SchemaURL != "" {
+		opts = append(opts, log.WithSchemaURL(c.scope.SchemaURL))
+	}
+	return c.provider.Logger(c.scope.Name, opts...)
+}
+
+// Option configures a [zapcore].
```
Comment (pellared) [inline bridges/otelzap/config.go:49]:
```suggestion
// Option configures a [Core].
```

File: bridges/otelzap/config.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+type Option interface {
+	apply(config) config
+}
+
+type optFunc func(config) config
+
+func (f optFunc) apply(c config) config { return f(c) }
+
+// WithInstrumentationScope returns an [Option] that configures the scope of
+// the [log.Logger] used by a [zapcore].
```
Comment (pellared) [inline bridges/otelzap/config.go:59]:
```suggestion
// the [log.Logger] used by a [Core].
```

File: bridges/otelzap/config.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+//
+// By default if this Option is not provided, the zapcore will use a default
```
Comment (pellared) [inline bridges/otelzap/config.go:61]:
```suggestion
// By default if this Option is not provided, [Core] will use a default
```

File: bridges/otelzap/core.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+// Core is a [zapcore.Core] that sends logging records to OpenTelemetry.
+type Core struct {
+	logger log.Logger
+}
+
+// Compile-time check *Core implements zapcore.Core.
+var _ zapcore.Core = (*Core)(nil)
+
+// NewOTelZapCore creates a new [zapcore.Core] that can be used with zap.New()
+// this instance will translate zap logs to opentelemetry logs and export them.
```
Comment (pellared) [inline bridges/otelzap/core.go:27]:
```suggestion
// NewCore creates a new [zapcore.Core] that can be used with [go.uber.org/zap.New].
```

File: bridges/otelzap/core.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+	return true
+}
+
+// TODO
+// With adds structured context to the Core.
+func (o *Core) With(fields []zapcore.Field) zapcore.Core {
+	return o
+}
+
+// Sync flushes buffered logs (if any).
```
Comment (pellared) [inline bridges/otelzap/core.go:47]: Add a TODO comment as I think Sync will be a no-op.

File: bridges/otelzap/core_test.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+// Copyright (c) 2016-2017 Uber Technologies, Inc.
```
Comment (pellared) [inline bridges/otelzap/core_test.go:4]: Is there anything taken from Uber code here?

```suggestion
```

File: versions.yaml [hunk 8]
```diff
@@ -81,3 +81,4 @@ excluded-modules:
   - go.opentelemetry.io/contrib/instrgen
   - go.opentelemetry.io/contrib/instrgen/driver
   - go.opentelemetry.io/contrib/instrgen/testdata/interface
+  - go.opentelemetry.io/contrib/bridges/otelzap
```
Comment (pellared) [inline versions.yaml:84]: Can you reorder (alphabetical) and add an EOL?

File: bridges/otelzap/core_test.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+				WithInstrumentationScope(scope),
+			)
+		})
+		assert.NotNil(t, h.logger)
+		require.IsType(t, &logtest.Recorder{}, h.logger)
+
+		l := h.logger.(*logtest.Recorder)
+		assert.Equal(t, scope.Name, l.Result()[0].Name)
+		assert.Equal(t, scope.Version, l.Result()[0].Version)
+		assert.Equal(t, scope.SchemaURL, l.Result()[0].SchemaURL)
```
Comment (pellared) [inline bridges/otelzap/core_test.go:48]:
```suggestion
		l := h.logger.(*logtest.Recorder)
		require.Len(t, l.Result(), 1)

		want := &logtest.ScopeRecords{Name: scope.Name, Version: scope.Version, SchemaURL: scope.SchemaURL}
		got := l.Result()[0]
		assert.Equal(t, want, got)
```

File: bridges/otelzap/core_test.go [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+		global.SetLoggerProvider(r)
+
+		var h *Core
+		assert.NotPanics(t, func() { h = NewCore() })
+		assert.NotNil(t, h.logger)
+		require.IsType(t, &logtest.Recorder{}, h.logger)
+
+		l := h.logger.(*logtest.Recorder)
+		want := &logtest.ScopeRecords{Name: bridgeName, Version: version}
+		assert.Equal(t, want, l.Result()[0])
```
Comment (pellared) [inline bridges/otelzap/core_test.go:29]:
```suggestion
		require.NotPanics(t, func() { h = NewCore() })
		require.NotNil(t, h.logger)
		require.IsType(t, &logtest.Recorder{}, h.logger)
		l := h.logger.(*logtest.Recorder)
		require.Len(t, l.Result(), 1)
		
		want := &logtest.ScopeRecords{Name: bridgeName, Version: version}
		got := l.Result()[0]
		assert.Equal(t, want, got)
```

**Comment by pellared** [comment]
Can you also update the `CODEOWNERS` file? You can add me as one of the codeowners.

Do you want to be a codeowner as well? Will you be able to maintain otelzap after your internship is finished?

**Comment by khushijain21** [comment]
yes! would like to maintain it after the internship too

**Review by MrAlias — APPROVED** [review]

**Review by pellared — APPROVED** [review]

**Comment by pellared** [comment]
@open-telemetry/go-approvers, PTAL. This PR needs a second approval from a non-Splunk employee.

**Review by dashpole — APPROVED** [review]

**Pull request merged**